library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
-    // |..........xxxxxxxxxx|xxxxxxxxxxxxxxxxxx....|......................| p2
-    // |....................|......................|......................|
-    // |.................c23|..c13.................|......................|
-    // |xxxxxxxxxxxxxxxxxxxx|xxxxx.................|......................| p3
-    // |....................|......................|......................|
-    // |.........c24........|......................|......................|
-    // |xxxxxxxxxxxx........|......................|......................| p4
+  def %(b: RuntimeLong): RuntimeLong = {
+    val alo = a.lo
+    val ahi = a.hi
+    val blo = b.lo
+    val bhi = b.hi
 
-    val c00 = p0 & 0x3fffff;
-    val c01 = (p1 & 0x1ff) << 13;
-    val c0 = c00 + c01;
+    if (isZero(blo, bhi))
+      throw new ArithmeticException("/ by zero")
 
-    val c10 = p0 >> 22;
-    val c11 = p1 >> 9;
-    val c12 = (p2 & 0x3ffff) << 4;
-    val c13 = (p3 & 0x1f) << 17;
-    val c1 = c10 + c11 + c12 + c13;
+    if (isInt32(alo, ahi)) {
+      if (isInt32(blo, bhi)) {
+        if (blo != -1) new RuntimeLong(alo % blo)
+        else Zero // Work around https://github.com/ariya/phantomjs/issues/12198
+      } else {
+        // Either a == Int.MinValue && b == (Int.MaxValue + 1), or (abs(b) > abs(a))
+        if (alo == Int.MinValue && (blo == 0x80000000 && bhi == 0)) Zero
+        else a // because abs(b) > abs(a)
+      }
+    } else {
+      val (aNeg, aAbsLo, aAbsHi) = inline_abs(alo, ahi)
+      val (bNeg, bAbsLo, bAbsHi) = inline_abs(blo, bhi)
+      val absR = unsigned_%(aAbsLo, aAbsHi, bAbsLo, bAbsHi)
+      if (aNeg) inlineLongUnary_-(absR.lo, absR.hi)
+      else absR
+    }
+  }
 
-    val c22 = p2 >> 18;
-    val c23 = p3 >> 5;
-    val c24 = (p4 & 0xfff) << 8;
-    val c2 = c22 + c23 + c24;
+  private def unsigned_%(alo: Int, ahi: Int, blo: Int, bhi: Int): RuntimeLong = {
+    if ((ahi & 0xffe00000) == 0) {
+      if ((bhi & 0xffe00000) == 0) {
+        val aDouble = ahi * TwoPow32 + alo.toUint
+        val bDouble = bhi * TwoPow32 + blo.toUint
+        val rDouble = aDouble % bDouble
+        new RuntimeLong(rawToInt(rDouble), rawToInt(rDouble / TwoPow32))
+      } else {
+        new RuntimeLong(alo, ahi) // because b > a
+      }
+    } else {
+      if (bhi == 0 && isPowerOfTwo_IKnowItsNot0(blo)) {
+        val pow = log2OfPowerOfTwo(blo)
+        new RuntimeLong(alo & ((1 << pow) - 1), 0)
+      } else if (blo == 0 && isPowerOfTwo_IKnowItsNot0(bhi)) {
+        val pow = log2OfPowerOfTwo(bhi)
+        new RuntimeLong(alo, ahi & ((1 << pow) - 1))
+      } else {
+        unsignedDivModHelper(alo, ahi, blo, bhi,
+            AskRemainder).asInstanceOf[RuntimeLong]
+      }
+    }
+  }
 
-    // Propagate high bits from c0 -> c1, c1 -> c2
-    val c1n = c1 + (c0 >> BITS)
+  private def unsignedDivModHelper(alo: Int, ahi: Int, blo: Int, bhi: Int,
+      ask: Int): RuntimeLong | js.Tuple4[Int, Int, Int, Int] = {
```

Comment:
Yes, unapply assignment is not yet sufficiently optimized, especially for specialized tuple types. It creates unnecessary temporary variables. Since this is a code-size- and performance critical file, I avoid it pretty much all the time.